src/plugin_discovery/find_plugin_specs.js

```diff
@@ -0,0 +1,134 @@
+  const find$ = Observable.merge(
+    ...config.get('plugins.paths').map(createPackAtPath$),
+    ...config.get('plugins.scanDirs').map(createPacksInDirectory$)
+  )
+    .share();
+
+  const extendConfig$ = find$
+    // get the specs for each found plugin pack
+    .mergeMap(({ pack }) => (
+      pack ? pack.getPluginSpecs() : []
+    ))
+    .mergeMap(async (spec) => {
+      // extend the config service with this plugin spec and
+      // collect its deprecations messages if some of its
+      // settings are outdated
+      const deprecations = [];
+      await extendConfigService(spec, config, settings, (message) => {
+        deprecations.push({ spec, message });
+      });
+
+      return {
+        spec,
+        deprecations,
+      };
+    })
+    // extend the config with all plugins before determining enabled status
+    .let(waitForComplete)
+    .map(({ spec, deprecations }) => {
+      const rightVersion = spec.isVersionCompatible(config.get('pkg.version'));
+      const enabled = rightVersion && spec.isEnabled(config);
+      return {
+        spec,
+        deprecations,
+        enabledSpecs: enabled ? [spec] : [],
+        disabledSpecs: enabled ? [] : [spec],
+        invalidVersionSpecs: rightVersion ? [] : [spec],
+      };
+    })
+    // determin which plugins are disabled before actually removing things from the config
+    .let(waitForComplete)
+    .do(result => {
+      for (const spec of result.disabledSpecs) {
+        disableConfigExtension(spec, config);
+      }
+    })
+    .share();
+
+  return {
+    // plugin packs found when searching configured paths
+    pack$: find$
+      .mergeMap(result => (
+        result.pack ? [result.pack] : []
+      )),
+
+    // errors caused by invalid directories of plugin directories
+    invalidDirectoryError$: find$
+      .mergeMap(result => (
+        isInvalidDirectoryError(result.error) ? [result.error] : []
+      )),
+
+    // errors caused by directories that we expected to be plugin but were invalid
+    invalidPackError$: find$
+      .mergeMap(result => (
+        isInvalidPackError(result.error) ? [result.error] : []
+      )),
+
+    // { spec, message } objects produces when transforming deprecated
+    // settings for a plugin spec
+    deprecation$: extendConfig$
+      .mergeMap(result => result.deprecations),
+
+    // the config service we extended with all of the plugin specs,
+    // only emitted once it is fully extended by all
+    extendedConfig$: extendConfig$
+      .ignoreElements()
+      .concat([config]),
+
+    // all enabled PluginSpec objects
+    spec$: extendConfig$
+      .mergeMap(result => result.enabledSpecs),
+
+    // all disabled PluginSpec objects
+    disabledSpecs$: extendConfig$
+      .mergeMap(result => result.disabledSpecs),
+
+    // all PluginSpec objects that were disabled because their version was incompatible
+    invalidVersionSpecs$: extendConfig$
+      .mergeMap(result => result.invalidVersionSpecs),
+  };
+}
```

Comment:
This is great 🎉 🍕 🌮 
I like the use of Observables here. It's easy to see how `extendConfig$` gets built up.

Reply:
I iterated a fair amount and am glad you like where I landed, thanks!